src/ripple/consensus/ConsensusProposal.h

```diff
@@ -0,0 +1,272 @@
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+#ifndef RIPPLE_CONSENSUS_ConsensusProposal_H_INCLUDED
+#define RIPPLE_CONSENSUS_ConsensusProposal_H_INCLUDED
+
+#include <cstdint>
+#include <ripple/json/json_value.h>
+#include <ripple/protocol/JsonFields.h>
+
+namespace ripple
+{
+/** Represents a proposed position taken during a round of consensus.
+
+    During consensus, peers seek agreement on a set of transactions to
+    apply to the prior ledger to generate the next ledger.  Each peer takes a
+    position on whether to include or exclude potential transactions.
+    The position on the set of transactions is proposed to its peers as an
+    instance of the ConsensusProposal class.
+
+    An instance of ConsensusProposal can be either our own proposal or one of
+    our peer's.
+
+    As consensus proceeds, peers may change their position on the
+    transactions.  Each proposal has an increasing sequence number that orders
+    the successive proposals.  A peer may also choose to leave consensus
+    indicated by sending a ConsensuProposal with sequence equal to `seqLeave.
+
+    @tparam NodeID_t Type used to uniquely identify nodes/peers
+    @tparam LedgerID_t Type used to uniquely identify ledgers
+    @tparam Position_t Type used to represent the position taken on transactions
+                       under consideration during this round of consensus
+    @tparam Time_t Type used to represent times related to consensus
+ */
+template <
+    class NodeID_t,
+    class LedgerID_t,
+    class Position_t,
+    class Time_t>
+class ConsensusProposal
+{
+private:
+    //< Sequence value when a peer initially joins consensus
+    static std::uint32_t const seqJoin = 0;
+
+    //< Sequence number when  a peer wants to bow out and leave consensus
+    static std::uint32_t const seqLeave = 0xffffffff;
+
+public:
+
+    using NodeID = NodeID_t;
+
+    /** (Peer) Constructor
+
+        Constructs a peer's consensus proposal
+
+        @param prevLedger The previous ledger this proposal is building on.
+        @param seq The sequence number of this proposal.
+        @param position The position taken on transactions in this round.
+        @param closeTime Position of when this ledger closed.
+        @param now Time when the proposal was taken.
+        @param nodeID ID of node/peer taking this position.
+    */
+    ConsensusProposal(
+        LedgerID_t const& prevLedger,
+        std::uint32_t seq,
+        Position_t const& position,
+        Time_t closeTime,
+        Time_t now,
+        NodeID_t const& nodeID)
+    : previousLedger_(prevLedger)
+    , position_(position)
+    , closeTime_(closeTime)
+    , time_(now)
+    , proposeSeq_(seq)
+    , nodeID_(nodeID)
+    {
+
+    }
+
+    /** (Self) Constructor
```

Comment:
I have a changeset that started those changes, but ended up shelving it.  The main reason was I couldn't decide on the best names to use.  I'll give it another go though because I definitely agree that these should be distinct types.